connection.go

```diff
@@ -1163,3 +1193,71 @@ func (conn *Connection) NewStream() (*Stream, error) {
 		Conn: conn,
 	}, nil
 }
+
+// loadProtocolInfo sends info about client protocol,
+// receives info about server protocol in response
+// and store in in connection serverProtocolInfo.
+func (conn *Connection) loadProtocolInfo(w *bufio.Writer, r *bufio.Reader) error {
+	var resp Response
+	var err error
+
+	err = conn.writeProtocolInfoRequest(w, ClientProtocolVersion, ClientFeatures)
+	if err != nil {
+		return err
+	}
+
+	resp, err = conn.readProtocolInfoResponse(r)
+	if err != nil {
+		tarantoolError, ok := err.(Error)
+		if ok && tarantoolError.Code == ErrUnknownRequestType {
+			// IPROTO_ID requests are not supported by server.
+			conn.serverProtocolVersion = ProtocolVersionUnsupported
+			conn.serverFeatures = []Feature{}
+
+			return nil
+		}
+
+		return err
+	}
+
+	if len(resp.Data) == 0 {
+		return fmt.Errorf("Unexpected response on protocol info exchange: no data")
+	}
+
+	serverProtocolInfo, ok := resp.Data[0].(protocolInfo)
+	if !ok {
+		return fmt.Errorf("Unexpected response on protocol info exchange: wrong data")
+	}
+	conn.serverProtocolVersion = serverProtocolInfo.version
+	conn.serverFeatures = serverProtocolInfo.features
+
+	return nil
+}
+
+// ServerProtocolVersion returns protocol version supported by
+// connected Tarantool server.
+// Since 1.10.0
+func (conn *Connection) ServerProtocolVersion() ProtocolVersion {
+	return conn.serverProtocolVersion
+}
+
+// ClientProtocolVersion returns protocol version supported by
+// Go connection client.
+// Since 1.10.0
+func (conn *Connection) ClientProtocolVersion() ProtocolVersion {
+	return ClientProtocolVersion
+}
+
+// DoesServerSupportFeature checks if expected feature
+// is supported by connected Tarantool server.
+// Since 1.10.0
+func (conn *Connection) DoesServerSupportFeature(feature Feature) bool {
+	return isFeatureSupported(feature, conn.serverFeatures)
+}
+
+// DoesClientSupportFeature checks if expected feature
+// is supported by Go connection client.
+// Since 1.10.0
+func (conn *Connection) DoesClientSupportFeature(feature Feature) bool {
+	return isFeatureSupported(feature, ClientFeatures)
+}
```

Comment:
Why not just check values from ClientFeatures? It looks like duplicate functionality.

Reply:
It would be a bit inconvenient to have asymmetrical getters for client and server info. [2]
Also it is possible that we would enable/disable some features for a connection, like MP_ERROR encoding mode.

Reply:
> Also it is possible that we would enable/disable some features for a connection, like MP_ERROR encoding mode, so it became connection-dependent.

Then we need to give this opportunity, but hide ClientProtocolVersion and ClientFeatures from the public API.

Reply:
Well, to be honest, I don't get how I should design this one.

There should be a set of features
- provided by user
- required by user

and I don't see any way to mix them. Moreover, I don't see any way to actually allow user to provide a set of features. For example, they may not provide (intentionally or non-intentionally) `TransactionFeature` flag, but client actually supports this feature and we even provide the API for it. Either we start to disallow to use feature considering feature flag set or we allow user to misinform the Tarantool server. (For now, it is hard to say if this may have any consequences.)

Same goes about protocol version. To be honest, it is rather hard to design a solution for a such strange thing.

For now it seems that
- 3 flags doesn't do anything
- 1 flag changes the behavior when both client and server provides it

and, to be honest, I've planned to enable/disable error extension flag the following way:
- enable if user exports `_ = github.com/tarantool/go-tarantool/error_extension`
- do not enable otherwise.

Reply:
So it seems that for now we decided to not provide a user handles for feature setup. Constants are hidden now